src/main/java/frc/utils/pathplannerutils/PathPlannerUtils.java

```diff
@@ -0,0 +1,54 @@
+package frc.utils.pathplannerutils;
+
+import com.pathplanner.lib.auto.AutoBuilder;
+import com.pathplanner.lib.auto.NamedCommands;
+import com.pathplanner.lib.commands.PathfindingCommand;
+import com.pathplanner.lib.path.GoalEndState;
+import com.pathplanner.lib.path.PathConstraints;
+import com.pathplanner.lib.path.PathPlannerPath;
+import com.pathplanner.lib.pathfinding.Pathfinding;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Translation2d;
+import edu.wpi.first.wpilibj2.command.Command;
+
+import java.util.List;
+
+public class PathPlannerUtils {
+
+    public static void startPathPlanner() {
+        makePathPlannerCompatibleWithAdvantageKit();
+        scheduleWarmup();
+    }
+
+    private static void makePathPlannerCompatibleWithAdvantageKit() {
+        Pathfinding.setPathfinder(new LocalADStarAK());
+    }
+
+    private static void scheduleWarmup() {
+        PathfindingCommand.warmupCommand().schedule();
+    }
+
+    public static void registerCommand(String commandName, Command command) {
+        NamedCommands.registerCommand(commandName, command);
+    }
+
+    //todo - dynamic obstacles
+
+    public static Command createOnTheFlyPathCommand(Pose2d currentPose, Pose2d targetPose, PathConstraints constraints) {
```

Comment:
You can also overload this function with some default path constraints

Reply:
What is the difference between AutoBuilder.pathfindToPose and this?

I can't use the pose estimator, this class is a util class that meant to help things that use path planner (swerve mostly) to use it but should not be dependent on normal robot code. The wrap with current pose should be in swerve command.

Reply:
You are using auto builder here anyway, so it does depend on swerve. If you really want it to be independent, you should create a new PathfindHolonomic and get all the parameters that you put in configureAutoBuilder.

Reply:
It is not the same dependency. I don't need the specific swerve on the robot for this to work I only need to configure the auto builder. 
Also you suggested to use  "AutoBuilder.pathfindToPose", what is the difference between "createOnTheFlyPathCommand" and what you suggested?

Reply:
> What is the difference between AutoBuilder.pathfindToPose and this?

The AutoBuilder function creates the bezier points in your chosen pathfinder (LocalADStar.createPathPoints). It's basically the same algorithm but I would argue that it's more correct to access it that way.

Reply:
> I don't need the specific swerve on the robot for this to work I only need to configure the auto builder.
I can make the same argument for AutoBuilder.pathfindToPose

Reply:
> > What is the difference between AutoBuilder.pathfindToPose and this?
> 
> The AutoBuilder function creates the bezier points in your chosen pathfinder (LocalADStar.createPathPoints). It's basically the same algorithm but I would argue that it's more correct to access it that way.

So why trigon doing if statement and switching between this two?

Reply:
> > I don't need the specific swerve on the robot for this to work I only need to configure the auto builder.
> 
> I can make the same argument for AutoBuilder.pathfindToPose

Right I don't have problem with entering "AutoBuilder.pathfindToPose" instead. I have problem with using current pose and not getting it as param.

Reply:
> So why trigon doing if statement and switching between this two?

idk ask @levyishai 
unless you understand their code, you shouldn't copy it

Reply:
> > So why trigon doing if statement and switching between this two?
> 
> idk ask @levyishai unless you understand their code, you shouldn't copy it

If they did something, they probably had a reason. It could be worth checking why they did what they did in order to avoid dealing with problems later.

Reply:
I'm assuming you're referring to [this](https://github.com/Programming-TRIGON/RobotCode2024/blob/05e8936ede6f60737328ba2df8b5f09430e05a26/src/main/java/frc/trigon/robot/subsystems/swerve/SwerveCommands.java#L153) if statement.
You might notice that PathPlanner has a [similar if statement](https://github.com/mjansen4857/pathplanner/blob/05a123d86f47976c564564674fdb997e25b03773/pathplannerlib/src/main/java/com/pathplanner/lib/commands/PathfindingCommand.java#L182) in their pathfind logic.
This basically means that if you're too close to the target position, nothing will happen. So if that's the case, we just use create and follow a normal path.

Reply:
For the difference between pathfinding to creating path points: it's that pathfinding will create the most optimal path while avoiding the obstacles in the navgrid, and pathpoints will just create a normal path from your points with no regard to the navgrid.

Reply:
> I'm assuming you're referring to [this](https://github.com/Programming-TRIGON/RobotCode2024/blob/05e8936ede6f60737328ba2df8b5f09430e05a26/src/main/java/frc/trigon/robot/subsystems/swerve/SwerveCommands.java#L153) if statement. You might notice that PathPlanner has a [similar if statement](https://github.com/mjansen4857/pathplanner/blob/05a123d86f47976c564564674fdb997e25b03773/pathplannerlib/src/main/java/com/pathplanner/lib/commands/PathfindingCommand.java#L182) in their pathfind logic. This basically means that if you're too close to the target position, nothing will happen. So if that's the case, we just use create and follow a normal path. Btw - I just noticed the number isn't exactly correct. And after discussing with PathPlanner's creator, he just made the command quit if we're too close to the end position, which can simplify the logic.

So what it the advantage of using the method in the if if close to target?
Also why do you mirror the target pose instead of using blue alliance pose?

Reply:
For the first question; 
0.5 meters isn't that close for our needs. Our tolerance is 0.05 meters, so we need some kind of movement towards there. We could just PID there, but it'll be slower probably, and we also do pid after following the path for more position accuracy.
For the pose flipping, imo, both ways are acceptable. Since we already need a way to flip for PID and other things, we chose to flip manually. It's just more consistent with our code and maybe a bit more readable in terms of knowing exactly what's happening. Probably for you doing it automatically with PathPlanner's lib is better. 
Btw, I wouldn't advise looking at main with all the `AlliancePose2d` stuff - after the season I figured it's just too ugly and hard to understand, so I implemented [Mirrorable objects](https://github.com/Programming-TRIGON/RobotCode2024/tree/better-shooting/src/main/java/frc/trigon/robot/utilities/mirrorable) in the `better-shooting` branch.

Reply:
So when you use the own creation it’s not stopping when close right?
Yea I saw, I preferred to still have option to red pose so I implemented alliance rotation and translation and used them in pose.
In your check in the if you check your blue pose and the target mirror, why u mirror only the target, cause like that it does not measure the real distance. So if I give the create on fly blue pose as current and target I should cancel the in the create on fly the prevent flipping right ?

Reply:
> Yea I saw, I preferred to still have option to red pose so I implemented alliance rotation and translation and used them in alliance pose. (I liked the alliance positions, so I decided to stick with them.)
imo, mirrorable objects will be better for you. They allow for also choosing weather or not to mirror on red alliance, which gives you more control and abstraction of actions. The main reason for switching to them is the one I mentioned above, and the fact that using AlliancePose2ds for the current pose isn't a good idea and makes the code hard to read. With mirrroable objects you can show the fact that you only need to mirror the target position.

> In your check in the if you check your blue pose and the target mirror, why u mirror only the target, cause like that it does not measure the real distance.
The idea is that everything should be relative to the blue alliance. Everything. 
So the current pose is always relative to the blue alliance when given from the pose estimator. But, for example,  when I want to drive to Amplifier, and I'm on the read alliance, I'll want to mirror the measured blue alliance amp pose to reach the red alliance amp pose relative to the blue alliance. Might be harder to understand at first.

> Also, I still can't understand all the mirrors, you can do all with blue alliance pose and it will work fine, and like that remove the "direction" variable in the pid function and probably some more things like that.
If I won't have all the mirror stuff, I wouldn't be able to reach say the Amplifier's pose when I'm on the red alliance, because it will point to the blue Amplifier and not the red Amplifier.
The reason for the `direction` variable is that the field relative heading does change when I'm on the red alliance. So positive field relative power decreases my pose, instead of increasing it.

P.S. Sorry for the late response, didn't notice the email.

Reply:
> > Yea I saw, I preferred to still have option to red pose so I implemented alliance rotation and translation and used them in alliance pose. (I liked the alliance positions, so I decided to stick with them.)
> 
> IMO, mirrorable objects will be better for you. They allow for also choosing weather or not to mirror on red alliance, which gives you more control and abstraction of actions. The main reason for switching to them is the one I mentioned above, and the fact that using AlliancePose2ds for the current pose isn't a good idea and makes the code hard to read. With mirrroable objects you can show the fact that you only need to mirror the target position.
> 
> > In your check in the if you check your blue pose and the target mirror, why u mirror only the target, cause like that it does not measure the real distance.
> 
> The idea is that everything should be relative to the blue alliance. Everything. So the current pose is always relative to the blue alliance when given from the pose estimator. But, for example, when I want to drive to Amplifier, and I'm on the read alliance, I'll want to mirror the measured blue alliance amp pose to reach the red alliance amp pose relative to the blue alliance. Might be harder to understand at first.
> 
> > Also, I still can't understand all the mirrors, you can do all with blue alliance pose and it will work fine, and like that remove the "direction" variable in the pid function and probably some more things like that.
> 
> If I won't have all the mirror stuff, I wouldn't be able to reach say the Amplifier's pose when I'm on the red alliance, because it will point to the blue Amplifier and not the red Amplifier. The reason for the `direction` variable is that the field relative heading does change when I'm on the red alliance. So positive field relative power decreases my pose, instead of increasing it.
> 
> P.S. Sorry for the late response, didn't notice the email.

Oh, I get it. I didn't realize you use mirror like that and thought you save all poses as blue, so like that the red speaker saved as 16,5 blue pose. It makes sense now thanks for the explanation.

When I look again because of the minor use of mirror and red alliance pose (just like field const and field relative drive), I agree the alliance pose add a bit more complexity to the code. 

Thanks for all the help! ⭐🫵

Reply:
Glad you understood it!
I highly advise just saving blue side elements and mirroring them, it's much easier.
Either way, if you need any more clarification about our code I'm always here!

Reply:
Thank you!
 About the pid to pose function, I don't realize why you need the direction if all in "blue pose", also when i run with direction it move to other side then expected. Also weird behavior I saw from pid to pose: When I ran the func as move to pose and not as correction to path planner, without direction to (5, 8, 90Degrees) it drove to it and then started doing spiral around it that only gets bigger and bigger. (all runs were in simulation)

Reply:
The need for direction is because when I'm in the red alliance, the field relative angle isn't 0 degrees, but 180 degrees. 
That means that if I'm giving positive x power, I'll go in the negative x direction, so I need to flip the direction if I'm on the red alliance. 
When you run it with direction (I'm assuming you're talking about only PID without pathfinding, right?), are giving the PID controller the blue relative poses? 
For the spiral, it might be just bad PID gains, try lowering them maybe? Are these issues only happening when you're on the red alliance?

Reply:
Thanks for all the help, sorry for all the questions, but i have another one. 
When I drive in simulation and spin while drive to robot deviating left or right (depends on turn direction). 
In your code does it do the same or the sim code should also fix that?

Reply:
> Thanks for all the help, sorry for all the questions, but i have another one.
When I drive in simulation and spin while drive to robot deviating left or right (depends on turn direction).
In your code does it do the same or the sim code should also fix that?
I know that there is the discretize, and then you own reduce skew but in sim is does not fix it completely in our code.

First of all, don't apologize for asking questions, it's my pleasure!
The reason discretize won't work in sim but will work irl is that the sim code runs much much slower, a lot of loop overruns, and overall not 0.02 seconds per loop. Discretize accepts a time when using it, and we apply 0.02, but because the loop time isn't exact in sim, it still doesn't work fully.

Reply:
Based on what you say doing "currentTime - lastTime;" when last time and current update every cycle should make it more accurate.
But not much and the result in sim is still kinda bad (in change of 16 in x, y changing by 5).

In the season we implemented our own discretize with different number than 1 to solve this, and also we had this problem on real robot so the changeable factor kind of dealed with it. This solution sounds a bit funny to me and probably not the right thing to do.  Also I didn't check already the swerve on carpet with a large place to run so I can't say if the problem continue on real. Do you have suggestion how to solve the problem because even if on real it is working it is problematic to check swerve is sim when its so different from real?

Reply:
Some more question came to my mind.

Why u used profile pid controller. I get that he is more accurate but also significant more slower that normal well calibrated pid controller?

Why did you cancel the reset pose by path planner auto builder?

In pose estimator, it starts all the wheel positions in 0 (new SwerveModuleState()). So first update can make a very big error if wheels didn't started at 0, You can see i solved it by saying take value in first update instead of doing pose estim calculation, but i didn't see at your code a solution for this problem. How did you handel it?

Reply:
> Based on what you say doing "currentTime - lastTime;" when last time and current update every cycle should make it more accurate.
But not much and the result in sim is still kinda bad (in change of 16 in x, y changing by 5).
I didn’t check this code on real robot on carpet, so i will check if the problem continue in real and update.

I'm confident this must be another problem than kinematics. Maybe bad velocity / position control on the modules? This is too much of a drift. Check how everything looks with Advantage Scope's swerve visualizer. 
Where's your newest swerve code? I'll take a look and see if I can find anything.

> Another thing about the swerve that is not related. How do i find coupling ratio and do i need to add the distance moved by coupling to the motor drive distance in inputs to make odometry more accurate? (I didn't saw that on code but it make sense)

We had coupling ratio compensation and still have IIRC. We just found out it really doesn't affect nearly enough to create so much more complexity in code (especially with 250hz odom).
The coupling ratio was like, very very small.
Either way, to check it, move the rotation wheel 360 degrees, see how much the drive has moved, and do 1/driveDiff. This can depend on how you use it in code exactly.

> Why u used profile pid controller. I get that he is more accurate but also significant more slower that normal well calibrated pid controller?

A profiled PID controller isn't slower. It should represent the fastest way to get to a point. When calibrated correctly, it's much faster than normal PID and has basically little-to-no overshot. And it's also safer.

> Why did you cancel the reset pose by path planner auto builder?

For us, the vision already resets the position, so we didn't need it. It just moves me a little from my real position if it resets (the robot isn't placed 100% correctly).

> In pose estimator, it starts all the wheel positions in 0 (new SwerveModuleState()). So first update can make a very big error if wheels didn't started at 0, You can see i solved it by saying take value in first update instead of doing pose estim calculation, but i didn't see at your code a solution for this problem. How did you handel it?

It doesn't really matter since the pose right when the code started is always wrong, so a little bit off from 0,0 doesn't really matter. Vision / PathPlanner will reset the pose. It won't affect the game.

Reply:
> > Based on what you say doing "currentTime - lastTime;" when last time and current update every cycle should make it more accurate.
> > But not much and the result in sim is still kinda bad (in change of 16 in x, y changing by 5).
> > I didn’t check this code on real robot on carpet, so i will check if the problem continue in real and update.
> 
> I'm confident this must be another problem than kinematics. Maybe bad velocity / position control on the modules? This is too much of a drift. Check how everything looks with Advantage Scope's swerve visualizer. Where's your newest swerve code? I'll take a look and see if I can find anything.
>
branch “working-trigon-swerve” there is open pr also
> > Another thing about the swerve that is not related. How do i find coupling ratio and do i need to add the distance moved by coupling to the motor drive distance in inputs to make odometry more accurate? (I didn't saw that on code but it make sense)
> 
> We had coupling ratio compensation and still have IIRC. We just found out it really doesn't affect nearly enough to create so much more complexity in code (especially with 250hz odom). The coupling ratio was like, very very small. Either way, to check it, move the rotation wheel 360 degrees, see how much the drive has moved, and do 1/driveDiff. This can depend on how you use it in code exactly.
> 
But when you move the rotation 360 the drive wheel moves but the drive motor not so how did you know how much it moved? 
> > Why u used profile pid controller. I get that he is more accurate but also significant more slower that normal well calibrated pid controller?
> 
> A profiled PID controller isn't slower. It should represent the fastest way to get to a point. When calibrated correctly, it's much faster than normal PID and has basically little-to-no overshot. And it's also safer.
> 
Oh ok, probably because I checked on sim on the values are not calibrated perfect.
> > Why did you cancel the reset pose by path planner auto builder?
> 
> For us, the vision already resets the position, so we didn't need it. It just moves me a little from my real position if it resets (the robot isn't placed 100% correctly).
> 
> > In pose estimator, it starts all the wheel positions in 0 (new SwerveModuleState()). So first update can make a very big error if wheels didn't started at 0, You can see i solved it by saying take value in first update instead of doing pose estim calculation, but i didn't see at your code a solution for this problem. How did you handel it?
> 
> It doesn't really matter since the pose right when the code started is always wrong, so a little bit off from 0,0 doesn't really matter. Vision / PathPlanner will reset the pose. It won't affect the game.

Reply:
> branch “working-trigon-swerve” there is open pr also

I don't think this is necessarily the problem, but I need to point this out. 
[This line](https://github.com/GreenBlitz/Off-Season-2024/blob/f58aeb75acb656417804e8f3706ea04099ca5c7e/src/main/java/frc/robot/subsystems/swerve/modules/simulationmodule/SimulationModule.java#L51) is very problematic. It defies the point of voltage compensation and creates inconsistencies. I don't think / know if this even works in sim, but don't do stuff like this. The whole idea of vcomp is that you specify ONE consistent number so your control will be consistent between batteries.
There are other unrelated things in the general code which should be changed. 
Like [using SmartDashboard here](https://github.com/GreenBlitz/Off-Season-2024/blob/f76fa84d43770a818bb42c25bfcb39e9b5ba8fca/src/main/java/frc/utils/utilcommands/DashboardCommand.java#L17) (which is incompatible which AdvantageKit) and other things I remember I've noticed. But it will require a more thorough review on my part. 

> But when you move the rotation 360 the drive wheel moves but the drive motor not so how did you know how much it moved?

Which motor / swerve module are you using? It should be detected. If not, there's either backlash (which I don't think there is since the coupling is very early on) or just a bad internal encoder.

> Weird, when I used profiled pid controller for the rotation it moves the swerve in 4 rad per sec but when I used pid controller with same pid values it moved in 11 rad per sec

Your profile probably just limits the speed too much. Give it more speed / accel.

Reply:
> > branch “working-trigon-swerve” there is open pr also
> 
> I don't think this is necessarily the problem, but I need to point this out. [This line](https://github.com/GreenBlitz/Off-Season-2024/blob/f58aeb75acb656417804e8f3706ea04099ca5c7e/src/main/java/frc/robot/subsystems/swerve/modules/simulationmodule/SimulationModule.java#L51) is very problematic. It defies the point of voltage compensation and creates inconsistencies. I don't think / know if this even works in sim, but don't do stuff like this. The whole idea of vcomp is that you specify ONE consistent number so your control will be consistent between batteries. There are other unrelated things in the general code which should be changed. Like [using SmartDashboard here](https://github.com/GreenBlitz/Off-Season-2024/blob/f76fa84d43770a818bb42c25bfcb39e9b5ba8fca/src/main/java/frc/utils/utilcommands/DashboardCommand.java#L17) (which is incompatible which AdvantageKit) and other things I remember I've noticed. But it will require a more thorough review on my part.
> 
but the vcomp should not be the maximum that can reach? so if battery low the max voltage is lower. also in sim the value that got from this func is 12 and only on real it depends on real battery.
about the dashboard Most of the times i would not care about what runs with this command, but you're right it is better to add dashboard logged value class to make it work with advKit. 
> > But when you move the rotation 360 the drive wheel moves but the drive motor not so how did you know how much it moved?
> 
> Which motor / swerve module are you using? It should be detected. If not, there's either backlash (which I don't think there is since the coupling is very early on) or just a bad internal encoder.

We are using mk4i (l3). As I understood with noam the steer spins the linear gear without spinning the motor so you can't follow the value from the motor. (when i tried it i rotated the steer 360 and the drive motor said he move 1 deg but real wheel did like 1/4 spin).
> 
> > Weird, when I used profiled pid controller for the rotation it moves the swerve in 4 rad per sec but when I used pid controller with same pid values it moved in 11 rad per sec
> 
> Your profile probably just limits the speed too much. Give it more speed / accel.

I gave it 10 rad per sec vel and acc and got to 5 rad per sec and with the pid to 11 with same pid values. I don't think this limits make such a big difference.

Reply:
> but the vcomp should not be the maximum that can reach? so if battery low the max voltage is lower. 

No. The vcomp should just be an arbitrary value that expresses about 100% voltage. But the idea is that it's constant and specific. You're basically going back to duty cycle with your method. Your values are very battery specific, and that's bad. 
I think [this post](https://www.chiefdelphi.com/t/voltage-compensation-vs-just-controlling-voltage/413761/3?u=levyishai) captures the main idea of vcomp and why it has to be constant. 


> also in sim the value that got from this func is 12 and only on real it depends on real battery.
So this could not be the problem with discretize, did you notice something else that could make the problem?

Yeah, I figured. 
I didn't really see anything that caught my sight. Maybe it is just bad gains?
Try plotting them with advantage scope.

> We are using mk4i (l3). As I understood with noam the steer spins the linear gear without spinning the motor so you can't follow the value from the motor. (when i tried it i rotated the steer 360 and the drive motor said he move 1 deg but real wheel did like 1/4 spin).

Do you have a neo for drive? Because a falcon/kraken would definitely notice that. In sds there's basically no backlash so it's improbable you're not seeing anything. Maybe it's because the neo has a bad encoder? 
Wdym be "but real wheel did like 1/4 spin". For 360 degrees of spin you got 1/4 rotations? That doesn't seem right. Maybe I misunderstood you.

> I gave it 10 rad per sec vel and acc and got to 5 rad per sec and with the pid to 11 with same pid values. I don't think this limits make such a big difference.

Maybe slow accel? 
Plot the target profiled angle vs current angle and see if you're following the profile correctly. Keep in mind you'll probably need more aggressive PID gains with a profile. 
If you are following it correctly, speed it up. 
If you are not following it correctly, adjust the gains. 
Do that until you reach the limit you can have.

Reply:
> > but the vcomp should not be the maximum that can reach? so if battery low the max voltage is lower.
> 
> No. The vcomp should just be an arbitrary value that expresses about 100% voltage. But the idea is that it's constant and specific. You're basically going back to duty cycle with your method. Your values are very battery specific, and that's bad. I think [this post](https://www.chiefdelphi.com/t/voltage-compensation-vs-just-controlling-voltage/413761/3?u=levyishai) captures the main idea of vcomp and why it has to be constant.
> 
> > also in sim the value that got from this func is 12 and only on real it depends on real battery.
> > So this could not be the problem with discretize, did you notice something else that could make the problem?
> 
> Yeah, I figured. I didn't really see anything that caught my sight. Maybe it is just bad gains? Try plotting them with advantage scope.

What i should plot and what i am suppose to see?
> 
> > We are using mk4i (l3). As I understood with noam the steer spins the linear gear without spinning the motor so you can't follow the value from the motor. (when i tried it i rotated the steer 360 and the drive motor said he move 1 deg but real wheel did like 1/4 spin).
> 
> Do you have a neo for drive? Because a falcon/kraken would definitely notice that. In sds there's basically no backlash so it's improbable you're not seeing anything. Maybe it's because the neo has a bad encoder? Wdym be "but real wheel did like 1/4 spin". For 360 degrees of spin you got 1/4 rotations? That doesn't seem right. Maybe I misunderstood you.

We have falcon500. I spin the steer by 360 deg and the linear pass about 1/6 (i exaggerate a bit) rotation but the motor itself says that he moved like 1 degree. Noam said this is how it suppose to be and the motor should not move because how the gearing is built. 
> 
> > I gave it 10 rad per sec vel and acc and got to 5 rad per sec and with the pid to 11 with same pid values. I don't think this limits make such a big difference.
> 
> Maybe slow accel? Plot the target profiled angle vs current angle and see if you're following the profile correctly. Keep in mind you'll probably need more aggressive PID gains with a profile. If you are following it correctly, speed it up. If you are not following it correctly, adjust the gains. Do that until you reach the limit you can have.

I don't get it, i give 15 rads in max vel and acceleration, and 12 p to the profile. and get little oscillations in 8 velocity of swerve. but with normal pid i get with 6 p 11 swerve vel and no osculation. Its like the profile is wrong.

Reply:
Also, why the u block the capacity of timestamps in the thread with 100? Like that after 100 values it cannot get more and return wrong time stamp for new updates.

Reply:
About the coupling, I calibrated it now. So apparently when you spin the steer without blocking the drive from spinning the motor isn't moving and same opposite. So I blocked the drive with a screw and rotated the steer 360 deg. the drive motor after gear ration calculation said it moved 0,59 rotations and this is the coupling ratio. When you spin the module you can see that the drive moves about 0.55 - 0.65 so the value make sense.

Reply:
About the coupling, I calibrated it now. So apparently when you spin the steer without blocking the drive from spinning the motor isn't moving and same opposite. So I blocked the drive with a screw and rotated the steer 360 deg. the drive motor after gear ration calculation said it moved 0,59 rotations and this is the coupling ratio. When you spin the module you can see that the drive moves about 0.55 - 0.65 so the value make sense.

Reply:
I found out the in sim I get with the modules and the swerve only to 5 meters per second and not 5.5. Do you know why is that? I checked and the modules gets 11.99 voltage and still can't pass 5.033 meters per second.

Also about discretize "solution" we found last season is to mult the time by factor but it depends on the vel of the robot and not really a good solution. you can see in the discretize draft pr.

Reply:
> What i should plot and what i am suppose to see? If you can it would be nice to do discord a call;

I'd love to get on a call whenever you want. 

> We have falcon500. I spin the steer by 360 deg and the linear pass about 1/6 (i exaggerate a bit) rotation but the motor itself says that he moved like 1 degree. Noam said this is how it suppose to be and the motor should not move because how the gearing is built.

This seems very weird to me. Could you maybe send a video showing the values from Phoenix tuner and the wheel moving? IIRC coupling ratio should go the other way, where the motor's encoder moves but not the actual wheel.

> I don't get it, i give 15 rads in max vel and acceleration, and 12 p to the profile. and get little oscillations in 8 velocity of swerve. but with normal pid i get with 6 p 11 swerve vel and no osculation. It’s like the profile is wrong.
Maybe it’s because of physics in sim so profile doesn’t have all his effect. I will check on real robot and update.

It's just a whole different way to calibrate from normal PID. You'll need much higher values. What exactly do you mean by oscillations? Could you send a link to the code? Keep in mind values will differ between sim to the real robot. 

>Another question I have, why you didn't used latency in odometry for signals?

When you update at 250hrtz there's not much room for latency. It's way too minimal for me to care :)

> About the coupling, I calibrated it now. So apparently when you spin the steer without blocking the drive from spinning the motor isn't moving and same opposite. So I blocked the drive with a screw and rotated the steer 360 deg. the drive motor after gear ration calculation said it moved 0,59 rotations and this is the coupling ratio. When you spin the module you can see that the drive moves about 0.55 - 0.65 so the value make sense.

This seems wayyyyy too much. I'd love to see a video 

> I found out the in sim I get with the modules and the swerve only to 5 meters per second and not 5.27 (max vel from sds website). Do you know why is that?@levyishai I checked and the modules gets 11.99 voltage and still can't pass 5.033 meters per second.

Values differ from sim to reality a lot. There are a lot of factors that are hard to take into consideration. It's probably just a wrong moment of inertia (`jKgMetersSquared`)

> Also about discretize, "solution" I found is using factor. You can see in the pr discretize the func. But it is kinda funny solution and i dont really understand why mult the time step helping.

Wdym by using a factor? Just a calibrated value? Could you send a video of the drift you mentioned before? 

Sorry for the very very late response, I was very busy with school stuff

Reply:
It’s fine thanks for the answeres. Can you do a call today when I’m in school with the robot. Like in the next 2 hours?

Reply:
In the next two hours I'll be in class unfortunately. 
I can join a call at around 6 if you're free

Reply:
Ok, I sent you friend request on discord (from the frc server). Call me or I will call around 6